Add docs for app defaulting and validation (#731)
giantswarm/docs: src/content/reference/app-defaulting-validation/index.md

---
title: Defaulting and Validation of App CRs
description: How defaulting and validation of app CRs is implemented by app-admission-controller
weight: 110
type: page
last_review_date: 2020-12-16
owner:
  - https://github.com/orgs/giantswarm/teams/team-batman
---

## Overview

> For Giant Swarm releases using app-operator version 3.0.0 and upwards the

Reviewer: ```suggestion
For Giant Swarm releases using app-operator version 3.0.0 and upwards (As of writing: Azure v13.1.0+, AWS and KVM TBD) the
```

Would it be helpful to specify which workload cluster releases has this as of today? Customers and SEs think in terms of those, right?

Author: Yes at the moment its just Azure >= v13.1.1 there are no AWS and KVM releases.

This is why it explains how to check if the release has app-operator v3.0.0. But I agree its easier for customers and SEs to state the versions. We just need to remember to update this.

 I'll add something.

Author: @cokiengchiara I added a note for this.

Reviewer: perfect!

> defaulting and validation logic of [App CRs](/reference/cp-k8s-api/apps.application.giantswarm.io/) is enabled. This logic is provided by

Reviewer 2: IMO using `relref` to spot dead links before deploying the site is still a good idea. I tested this locally, and this time the syntax and reference is correct ;)

```suggestion
defaulting and validation logic of [App CRs]({{< relref "/content/reference/cp-k8s-api/apps.application.giantswarm.io.md" >}}) is enabled. This logic is provided by
```

Author: Fine with me :)

Author: ```
ERROR 2021/01/18 13:15:29 [en] REF_NOT_FOUND: Ref "/content/reference/cp-k8s-api/apps.application.giantswarm.io.md": "/docs/src/content/reference/app-defaulting-validation/index.md:14:46": page not found
Error: Error building site: logged 1 error(s)
```

https://app.circleci.com/pipelines/github/giantswarm/docs/1930/workflows/97999bd9-43ed-4ef7-b353-fd7b552f7d21/jobs/4640

@ubergesundheit The build failed when I updated the branch. So I've reverted back to the previous link. Do you know why it failed?

Reviewer 2: Sorry for breaking your build (again) :(

Not your fault, because the master branch moved the document to `/content/reference/management-api/apps.application.giantswarm.io.md`.

So the build error was good after all to show that the old link pointing to `/reference/cp-k8s-api/apps.application.giantswarm.io/` was not correct any more.

So this suggestion should work (and break builds for people in the future who move the linked document)

```suggestion
defaulting and validation logic of [App CRs]({{< relref "/content/reference/management-api/apps.application.giantswarm.io.md" >}}) is enabled. This logic is provided by
```

Author: > Sorry for breaking your build (again) :(

Thanks @ubergesundheit I'd missed the renaming. Great that the build caught it!

[app-admission-controller](https://github.com/giantswarm/app-admission-controller).
You can check if your cluster has this version of app-operator by checking the
release details in the web interface.

> We have not enabled the defaulting and validation for existing app CRs to avoid

Reviewer 3: ```suggestion
We have not enabled the defaulting and validation for existing App CRs to avoid
```

disrupting your current usage of the App Platform.

## Defaulting

The defaulting logic is implemented using a [mutating admission webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook). It currently
defaults the following settings in the App CR if they are not specified.

> - The `app-operator.giantswarm.io/version` label that determines which instance

Reviewer: ```suggestion
- The `app-operator.giantswarm.io/version` label - It determines which instance
```

Author: Yes that's nicer thx!

of app-operator processes the App CR. The value is set based on the release
version of the cluster. e.g. `3.0.0`

> - The cluster values configmap that contains per cluster values like the base

Reviewer: ```suggestion
- The cluster values configmap - It contains per cluster values like the base
```

Reviewer: Minor suggestion to separate the thing being defaulted and what the thing does. 

It tripped me up, "Where is the verb??" then I remembered this is a continuation of the sentence, "It defaults the ff settings..."

domain for your cluster. e.g. `x7jwz-cluster-values`

> - The kubeconfig secret that lets app-operator connect to your cluster. e.g. `x7jwz-kubeconfig`

Reviewer: ```suggestion
- The kubeconfig secret - It lets app-operator connect to your cluster. e.g. `x7jwz-kubeconfig`
```

  name: kong-app
  namespace: kong
  version: 0.7.2
  userConfig:
    configMap:
      name: kong-user-values
      namespace: x7jwz
```

Here is the App CR with the defaults added by the mutating webhook.

```yaml
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  name: my-kong
  namespace: x7jwz
  labels:
    app-operator.giantswarm.io/version: 3.0.0
spec:
  catalog: giantswarm
  name: kong-app
  namespace: kong
  version: 0.7.2
  config:
    configMap:
      name: x7jwz-cluster-values
      namespace: x7jwz
  kubeConfig:
    context:
      name: x7jwz
     inCluster: false
     secret:
      name: x7jwz-kubeconfig
      namespace: x7jwz
  userConfig:
    configMap:
      name: kong-user-values
      namespace: x7jwz
```

## Validation

The validation logic is implemented using a [validating admission webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook).
It checks for common problems with App CRs.

Currently we validate:

- The `app-operator.giantswarm.io/version` label is present.
- All referenced configmaps and secrets exist.
- The catalog has a matching AppCatalog CR.

## Retry Logic

> During cluster creation there can a short delay while the kubeconfig

Reviewer: ```suggestion
During cluster creation there can be a short delay while the kubeconfig
```

secret and cluster values configmap are generated.

So if you have an automated process for creating App CRs please ensure

> it retries and the App CRs will be created once these resources exist.

Reviewer: Any guidance needed to do this?

---
title: Defaulting and Validation of App CRs
description: How defaulting and validation of app CRs is implemented by app-admission-controller
weight: 110
type: page
last_review_date: 2020-12-16
owner:
  - https://github.com/orgs/giantswarm/teams/team-batman
user_questions:
  - Is App CR defaulting and validation logic enabled for my cluster? 
  - How can I use App CR defaulting logic when installing Managed Apps?

>   - What fields are validated in App CRs when installing or updating Managed Apps?

Author: > I would really appreciate if search questions be added to the front matter. This will help in our overall effort in making docs more searchable and accessible.

@Oshratn Yes good point. I've added these questions.

---

> title: Defaulting and Validation of app CRs

Reviewer 4: ```suggestion
title: Defaulting and Validation of App CRs
```

description: How defaulting and validation of app CRs is implemented by app-admission-controller
weight: 110
type: page
last_review_date: 2020-12-16
owner:
  - https://github.com/orgs/giantswarm/teams/team-batman
---

## Overview

For Giant Swarm releases using app-operator version 3.0.0 and upwards the

> defaulting and validation logic is enabled. This logic is provided by

Reviewer 2: ```suggestion
defaulting and validation logic of [app CRs]({{< relref "/src/content/reference/cp-k8s-api/apps.application.giantswarm.io.md" >}} is enabled. This logic is provided by
```

Reviewer 2: Sorry for breaking the build.. I'll see whats the correct link

> app-admission-controller.

Reviewer 2: ```suggestion
[app-admission-controller](https://github.com/giantswarm/app-admission-controller).
```

> defaulting and validation logic of [app CRs]({{< relref "/content/reference/cp-k8s-api/apps.application.giantswarm.io.md" >}}) is enabled. This logic is provided by

Reviewer 4: We use this format throughout our docs:

```suggestion
defaulting and validation logic of [App CRs](/content/reference/cp-k8s-api/apps.application.giantswarm.io.md) is enabled. This logic is provided by
```

Also changed to use mixed case "App" for the CRD kind.

> defaulting and validation logic of [app CRs]({{< relref "/src/content/reference/cp-k8s-api/apps.application.giantswarm.io.md" >}} is enabled. This logic is provided by

Reviewer 2: This one should work

```suggestion
defaulting and validation logic of [app CRs]({{< relref "/content/reference/cp-k8s-api/apps.application.giantswarm.io.md" >}}) is enabled. This logic is provided by
```

Author: @ubergesundheit Thanks for checking. First time I've used relref and its very handy. 👍

> [app-admission-controller](https://github.com/giantswarm/app-admission-controller).

Reviewer 4: As a reader, I'd like to understand how I can learn which releases use app-operator v3.0.0 or higher. So we can either name them here, or explain where this can be found out (e. g. by checking the release details in the web interface.)

Author: Good point thanks. I'll explain how to check in the release details as I'm not sure which releases it will be in yet.

We have not enabled the defaulting and validation for existing app CRs to avoid

> disrupting your existing usage of App Platform.

Reviewer 4: The missing "the" in "usage of App Platform" sounds a bit strange to me. But I'm not the native speaker. Just pointing it out for alignment.

Author: I think it reads better as the "the App Platform" I'll fix.

## Defaulting

> The defaulting logic is implemented using a mutating webhook. It currently

Reviewer 4: ```suggestion
The defaulting logic is implemented using a [mutating admission webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook). It currently
```

defaults the following settings in the app CR if they are not specified.

- The `app-operator.giantswarm.io/version` label that determines which instance of app-operator processes the app CR.
- The kubeconfig secret that lets app-operator connect to your cluster.
- The cluster values configmap that contains per cluster values like the base

> domain for your cluster.

Reviewer 2: IMO it would be very useful to write down the actual default values (if possible) for these points

Author: Good idea. They will differ because the names have the cluster ID. But I'll add examples.

Here is an example app CR with only the settings you need to provide. The user
values configmap is optional, if you wish to configure the app with your own values.

```yaml
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  name: "my-kong"
  namespace: "x7jwz"
spec:
  catalog: "giantswarm"
  name: "kong-app"
  namespace: "kong"
  version: "0.7.2"
  userConfig:
    configMap:
      name: "kong-user-values"
      namespace: "x7jwz"

> ```

Reviewer 4: IMO both YAML snippets could be easier to read if the string quoting would be removed.

Author: Much nicer. I'll get rid of the quotes.

Here is the app CR with the defaults added by the mutating webhook.

```yaml
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  name: "my-kong"
  namespace: "x7jwz"
  labels:
    app-operator.giantswarm.io/version: "3.0.0"
spec:
  catalog: "giantswarm"
  name: "kong-app"
  namespace: "kong"
  version: "0.7.2"
  config:
    configMap:
      name: "x7jwz-cluster-values"
      namespace: "x7jwz"
  kubeConfig:
    context:
      name: "x7jwz"
     inCluster: false
     secret:
      name: "x7jwz-kubeconfig"
      namespace: "x7jwz"
  userConfig:
    configMap:
      name: "kong-user-values"
      namespace: "x7jwz"
```

## Validation

The validation logic is implemented using a validating webhook. It checks for
common problems with app CRs.

> Currently we validate.

Reviewer 2: ```suggestion
Currently we validate:
```

> The validation logic is implemented using a validating webhook. It checks for

Reviewer 4: ```suggestion
The validation logic is implemented using a [validating admission webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook). It checks for
```

- The `app-operator.giantswarm.io/version` label is present.
- All referenced configmaps and secrets exist.

> - The catalog has a matching appcatalog CR.

Reviewer 4: ```suggestion
- The catalog has a matching AppCatalog CR.
```

## Retry Logic

During cluster creation there can a short delay while the kubeconfig
secret and cluster values configmap are generated.

> So if you have an automated process for creating app CRs please ensure

Reviewer 4: ```suggestion
So if you have an automated process for creating App CRs please ensure
```

> it retries and the app CRs will be created once these resources exist.

Reviewer 4: ```suggestion
it retries and the App CRs will be created once these resources exist.
```
